_posts/2019-01-24-how-to-keep-safe-as-a-client.md

```diff
@@ -0,0 +1,129 @@
+}else{
+  errorCount ++;
+  if(errorCount >= limitCount){
+    isOpenCircuitBreaker = true;
+  }
+}
+
+```
+
+### Cut off connection
+
+Once our downstream system is broken, we need to cut off connection instantly. This action can both reduce useless remote call and reduce downstream system's load.
+For example, we use RPC framework to communicate between systems.
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vS47MgWD0aWMEqAYpvxibphTppSvBia6tRo3wpDgBgTKys8MqtgzwcqWXw/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+ServiceA is a client, If we find ServiceB is unavailable, we can return failed response within ServiceA and do not invoke ServiceB
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vS7HACeboGiaUlzjRXWJhR2nGMJQIYkfAxZE1TRfTIXME68kC6YPmRaIg/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+This is so called "Fail Fast". We can use below code to implement fail fast.
+
+```python
+if(isOpenCircuitBreaker == true){
+  return fail;
+}
+
+//Invoke remote service here
+
+```
+### Try to reconnect
+
+After cut off connections, we need to recover as soon as possible. Because currently the system is not complete. In order to automatically recover connection we need to define a standard, like previously how we judge a system is unavailable, we need to define whether a system is available again.
+- Threshold: for example, there are 100 success requests within 5 seconds.
+- Percentage: for example, there are 95% success requests within 10 seconds.
+
+It also contains time window, threshold and percentage. But there is a slight difference: most of the time, the system will keep unavailable for a moment, So we can check in a fixed times(30s for example).
+
+```python
+successCount = 0;
+isHalfOpen = true;
+
+if(success){
+  if(isHalfOpen){
+    successCount ++;
+    if(successCount == threshold){
+      isOpenCircuitBreaker = false;
+    }
+  }
+  return success;
+}else{
+  errorCount ++;
+  if(errorCount == threshold){
+    isOpenCircuitBreaker = true;
+  }
+}
+
+```
+However reconnect may failed, in a large system we can't use all servers to try reconnecting. We can take these measures:
+- 1.Use a few servers to check
+- 2.Add a health check interface to check system conditions: CPU, IO etc.
+
+### recover
+
+Once the system is available, we need to recover.
+
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vSUGPwicRFU6xJeNPxzKOWEhlM3yKSVdEfk1nusm5FfnRm1WyYkOVybBg/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+
+This is the general process of Circuit Breaker, frameworks like hystrix, resilience4j are all have such process flow.
+
+## Best practice for Circuit Breaker
+
+### What kind of situation require Circuit Breaker
+
+Circuit Breaker is not suitable in every situations. There are some conditions we can consider using Circuit Breaker:
+- The system we rely on is a shared system, and our client is just one of them. Other systems may abuse the shared system and affect us.
```

Comment:
Let me rephrase this a bit:
- If the target system is a shared service: when other systems abuse this shared service, our client can be affected.